Src/Support/Google.Apis/Upload/ResumableUpload.cs

```diff
@@ -690,40 +696,48 @@ private void MediaCompleted(HttpResponseMessage response)
         }
 
         /// <summary>Prepares the given request with the next chunk in case the steam length is unknown.</summary>
-        private async Task<UploadBuffer> PrepareNextChunkUnknownSizeAsync(Stream stream, CancellationToken cancellationToken)
+        private async Task PrepareBufferUnknownSizeAsync(Stream stream, CancellationToken cancellationToken)
         {
-            if (LastMediaBuffer == null)
+            // On the first iteration, create a new buffer. On later iterations, reuse any data that hasn't
+            // been successfully received yet.
+            if (Buffer is null)
             {
                 // Initialise state
                 // ChunkSize + 1 to give room for one extra byte for end-of-stream checking
-                LastMediaBuffer = new UploadBuffer(ChunkSize + 1);
+                Buffer = new UploadBuffer(this, ChunkSize + 1);
             }
             else
             {
-                LastMediaBuffer.MoveUnsentDataToStartOfBuffer(BytesClientSent, BytesServerReceived, ChunkSize);
+                Buffer.MoveUnsentDataToStartOfBuffer(BytesClientSent, BytesServerReceived);
             }
             // Read any more required bytes from stream, to form the next chunk.
             // We don't rely on reading StreamLength to determine whether we've finished reading or not, as
             // there are corner cases where it can be not unknown, even though we're in the "unknown size" case -
             // see https://github.com/googleapis/google-api-dotnet-client/issues/1449.
-            bool finished = await LastMediaBuffer.PopulateFromStreamAsync(stream, BufferSize, cancellationToken).ConfigureAwait(false);
+            bool finished = await Buffer.PopulateFromStreamAsync(stream, cancellationToken).ConfigureAwait(false);
             if (finished)
             {
-                StreamLength = BytesServerReceived + LastMediaBuffer.Length;
+                StreamLength = BytesServerReceived + Buffer.Length;
             }
-            return LastMediaBuffer;
         }
 
         /// <summary>Prepares the given request with the next chunk in case the steam length is known.</summary>
-        private async Task<UploadBuffer> PrepareNextChunkKnownSizeAsync(Stream stream, CancellationToken cancellationToken)
+        private async Task PrepareBufferKnownSizeAsync(Stream stream, CancellationToken cancellationToken)
         {
-            int chunkSize = (int)Math.Min(StreamLength - BytesServerReceived, (long)ChunkSize);
+            // On the first iteration, create a new buffer. On later iterations, just reset it so we can reuse the
+            // blocks inside. On the last iteration this may be larger than we need it to be, but that's okay -
+            // we just won't use it all.
+            if (Buffer is null)
+            {
+                Buffer = new UploadBuffer(this, chunkSize);
+            }
+            else
+            {
+                Buffer.Reset();
```

Comment:
Does the 308 response code can only happen when the original stream is not seekable? Why?
I mean, shouldn't this be `MoveUnsentDataToStartOfBuffer` as well?

Reply:
While we *could* do that, this path handles that issue in the block below, by seeking back to the right place instead.

Reply:
Ah true, didn't see that, sorry.

If the buffer were to be circular, you could saved the copying and the reseeking (could look like `Reset(BytesClientSent, BytesClientReceived)` for both seekable and unseekable), although if 308 doesn't happen that often, it might no be worth it.